src/hotspot/share/oops/instanceKlass.cpp

```diff
@@ -537,6 +537,8 @@ void InstanceKlass::deallocate_methods(ClassLoaderData* loader_data,
       // Only want to delete methods that are not executing for RedefineClasses.
       // The previous version will point to them so they're not totally dangling
       assert (!method->on_stack(), "shouldn't be called with methods on stack");
+      // Do the pointer maintenance before releasing the metadata
+      method->clear_jmethod_id();
```

Comment:
This causes a linear search through the method array of the ik for every method. Seeing that this is a workaround for a special case (an app that does a lot of retransforms *and* uses async profiler), I'd opt for making it conditional.

Reply:
Sadly, this is not async-profiler specific. The same issue can be observed by JVMTI only code grabbing a stacktrace.
What do you mean exactly by 'conditional'? Introducing a new JVM flag or something else?

Reply:
Ok, I see now - I could do the jmethodID maintenance only from `purge_previous_version_list()` call, leaving the proper metaspace deallocation untouched, therefore not adding unnecessary overhead there.

Reply:
I have modified the code to do jmethodID cleanup only when in `purge_previous_version_list()` - this should help with the added overhead.